### PR TITLE
Abstract the database

### LINKED ISSUE #13: Abstract Database
Current implementation uses dynamoDB, we should abstract the DB to something like postgres or mysql.

### PR DESCRIPTION
**Pull request by alessandromazza98**
Should close #13 
This is just the draft so that you can check

### CODE REVIEW

**Comment by 0xprames** [comment]
lmk when you've rebased against master and i'll take a closer look

**Comment by alessandromazza98** [comment]
Just to add some notes. What I plan immediately after merging this PR is:
- delete the `--save-to-json` flag that is now completely useless.
- delete also the `InMemoryDb` done using `HashMap`. We can replace it using an in memory db through SQLite or simply not adding it since the SQLite is already serverless and does not require any configuration to be run.
- update `README`.

**Comment by 0xprames** [comment]
can we split this up into

- SQLite db
- bringing the APIserver in (this is higher priority rn)

as well as a rebase?

**Comment by 0xprames** [comment]
i can take up the apiserver refactor - after we get that in, we can take a look at getting this sqlite PR in

File: bins/reth-crawler/src/api_server/peerdb/app_state.rs [hunk 1]
```diff
@@ -0,0 +1,23 @@
+use std::sync::Arc;
+
+use axum::extract::FromRef;
+use reth_crawler_db::{db::SqlPeerDB, AwsPeerDB, PeerDB};
+
+#[derive(Clone, FromRef)]
+pub struct AppState {
+    store: Arc<dyn PeerDB>,
```
Comment (0xprames) [inline bins/reth-crawler/src/api_server/peerdb/app_state.rs:8]: should be possible to make `AppState` here (and the rest of the references) generic over the PeerDB trait

**Review by 0xprames — APPROVED** [review]

**Comment by 0xprames** [comment]
i think overall this looks OK - im indexing a bit more on getting this in to unblock #28 more than anything else rn.

i would rather us use generics over the `dyn PeerDB` - but we can tackle that as a follow up

File: bins/reth-crawler/src/api_server/mod.rs [hunk 2]
```diff
@@ -1,18 +1,21 @@
...
 use axum::Json;
 use axum::Router;
-use dynamodb::{rest_router, AppState};
+use peerdb::{rest_router, AppState};
 use std::net::SocketAddr;
 use tracing::info;
 
-#[tokio::main]
-async fn main() -> Result<(), Box<dyn std::error::Error>> {
+pub async fn start_api_server(sql_db: bool) -> Result<(), Box<dyn std::error::Error>> {
```
Comment (0xprames) [inline bins/reth-crawler/src/api_server/mod.rs:10]: longer term - apiserver always starts with sqldb imo

File: bins/reth-crawler/src/crawler/factory.rs [hunk 3]
```diff
@@ -49,7 +49,7 @@ impl CrawlerFactory {
         }
     }
 
-    pub async fn make(&self) -> CrawlerService {
-        CrawlerService::new(self.discv4.clone(), self.dnsdisc.clone(), self.key).await
+    pub async fn make(&self, sql_db: bool) -> CrawlerService {
```
Comment (0xprames) [inline bins/reth-crawler/src/crawler/factory.rs:52]: eventually would be nice to remove the use of bool flags like this (`sql_db` works for now but in the future if there are more than 2 possibilities?)

probably makes sense to use an enum - and even then it'd be nice to see if there's a better way to do this (without passing in this flag at all, maybe creating the DB separately and passing it as a generic param to make?)

**Review by 0xprames — CHANGES_REQUESTED** [review]
should be ready for merge this after splitting up the DB part from apiserver part and only including the DB related changes

**Pull request closed without merging**